### ISSUE TITLE
Code error in attention part

### ISSUE THREAD

**Issue by HuangWei93** [PetarV-/GAT#23]
In my opinion, the coefficients part in your layer.py file should be coded in this way:

coefs = tf.exp(tf.nn.leaky_relu(logits))
coefs = coefs * tf.expand_dims(tf.sparse.to_dense(self.A),0)
coefs = coefs / tf.reduce_sum(coefs, -1,keepdims=True)
coefs = tf.nn.dropout(coefs,  self.keep_prob)

Maybe your implementation is wrong, because exp(0) = 1.

**Comment by PetarV-** [comment]
Hello,

Thanks for your comment! However, I believe that the code should be fine as-is, for both the sparse and dense layer.

If your comment concerned the dense layer, note that `bias_mat` is added to the matrix, making the non-existing entries have a coefficient which is a large negative number, causing their normalised coefficient to tend to zero.

If it regards the sparse layer, check the documentation for `tf.sparse_softmax` at:

https://www.tensorflow.org/api_docs/python/tf/sparse/softmax

as can be seen, it ignores the masked entries when computing the softmax. 

(Also, it should be noted that, in this case, `tf.sparse.to_dense` could create a potentially massive matrix and cause an OOM issue.)

**Comment by HuangWei93** [comment]
Thanks for your explanation!


Best regard!


Wei

________________________________
